# Домашнее задание №3

## Настройка среды

In [1]:
from huggingface_hub import notebook_login

notebook_login() # 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

## Глобальные параметры работы блокнота

In [2]:
# Выбранная контрольная точка модели
model_checkpoint = "google-t5/t5-small"

## Задание

[Ссылка](https://tn-gvu.mckx.ru/c/c4YcAACAurcBAEAA/sE43BQ/ZYcLPOYT_IPW6LK3/?u=https%3A%2F%2Fcontest.yandex.ru%2Fcontest%2F67637%2Fenter%2F%3Futm_source%3Dmindbox%26utm_medium%3Demail%26utm_campaign%3Dtraining6%26utm_content%3Ddigest%26utm_term%3D06.11.2024) на контест с заданием.

Необходимо обучить модель перевода с языка зетан на английский.

***Данные:*** https://disk.yandex.ru/d/u8mmcUBn64p_Nw

***Формат решения:*** json-lines в формате, аналогичной валидации, с переводами исходных текстов в тестовой выборке

***Метрика качества:*** BLEU между переводами и английскими референсами на закрытом тесте

## Загрузка данных

In [ ]:
# ! wget https://disk.yandex.ru/d/u8mmcUBn64p_Nw

In [94]:
from datasets import load_dataset, Features, Value

data_files = {
    "train": "train.jsonl",
    "validation": "val.jsonl",
    "test": "test_no_reference.jsonl"
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

dataset = load_dataset("json", data_files=data_files, features=features)

Посмотрим на один пример данных из каждой части датасета:

In [92]:
dataset['train'][4]

{'src': '◈◠ ◧▱◠▦ ◀◫◓ ▨◠◉ ◂▱◠▽◈◠▦ ◀◠▷◞◪◈◗◳◧◓■ ◉◧◐▾▦▱◨◐▾ ○▱◎◠▦▱◠◓◈◠▦ ▨◠◉◠▦ ▽◠▷◨◈◫▱▴◓▵',
 'dst': "He's talking about a few right here in Lisbon, mostly Jews who have escaped the Germans, that's all."}

In [93]:
dataset['validation'][4]

{'src': '◲◒▱▴▫◗▱◪▦ ▻▾◀ ◚◪ ◀◠◓▱◠◓◈◠ ◀◨ ◠◳ ◳◫◳▴▼◪▨ ◞◠▫◬◒▱◠◓▪ ▽◭▢◈◪ ◭◉ ◈▩◒▴◓▨◪▦ ◗◉▨◫ ◞◠▫◬◒▱◠◓▪ ◳▩▢◈▴ 6■6 ◠◓▫▫▪▵"',
 'dst': 'Sales of drinks in pubs and bars increased by 6.6% over the month, while sales of food increased by 3%."'}

Немного узнать об этом чудном языке не помешает. Посмотрим сколько уникальных символов в этом языке:

In [7]:
# Функция для подсчета уникальных символов
def count_unique_characters(dataset, feature_name):
    unique_characters = set()
    for split in dataset.keys():
        for example in dataset[split][feature_name]:
            if example is not None:
                unique_characters.update(example)
    return len(unique_characters), unique_characters

# Подсчет уникальных символов во всех частях датасета
num_unique_chars, unique_chars = count_unique_characters(dataset, "src")

print(f"Количество уникальных символов: {num_unique_chars}")
print(f"Уникальные символы: {unique_chars}")

Количество уникальных символов: 182
Уникальные символы: {'º', '▰', '▷', 'Î', '◰', 'ο', '\u200b', '◱', 'ð', 'þ', '◯', '°', '◤', ';', '`', '£', '▴', '▸', 'Ý', '◠', '\x9d', '▣', 'ï', ':', '@', '◖', '◮', '─', '▱', '=', 'ñ', '◳', '◂', '◃', '♪', '“', 'É', '¶', '#', '◡', '\xa0', 'á', '▫', '"', 'û', '•', '◒', 'é', 'ý', '4', '◔', '▵', '7', 'ø', 'ι', '▢', '▩', '/', '◭', '◊', '▿', '◉', '0', '◘', '◥', '▶', '▯', '▲', '\u202d', '◜', '◣', '▬', '—', '\\', '◅', '◝', '”', '◇', 'Â', '△', '&', '▦', '^', '◑', 'Þ', '◫', '◧', '◬', '▪', '1', '}', 'ă', '◚', 'ë', 'ú', '3', 'ó', '◛', 'å', '‚', '◍', '◌', '¿', 'ª', '?', '5', '§', 'ţ', '9', 'ä', '8', '6', '◪', '◨', "'", '¡', 'đ', '~', '▮', '◗', '◦', '▽', 'â', '_', '◢', 'è', 'ß', 'ν', '►', '(', '♫', '◓', '▨', '◆', '◲', '‘', '\x99', '\x9e', '▻', '’', '◈', '▼', '{', '◐', 'Ä', '2', '◄', '¤', ')', '○', '™', 'î', '▧', '◕', '●', '◞', '▥', '◀', '▭', '◙', 'í', '■', '[', 'ô', '–', '!', '$', '◎', '*', '´', '◩', '◁', '□', '▾', '◟', ']', '%', '+', ' ', '▹', '▤', '½'}


Посмотрим аналогичную статистику для целевых последовательностей:

In [8]:
# Подсчет уникальных символов во всех частях датасета для целевых последовательностей
num_unique_chars_dst, unique_chars_dst = count_unique_characters(dataset, "dst")

print(f"Количество уникальных символов в целевых последовательностях: {num_unique_chars_dst}")
print(f"Уникальные символы в целевых последовательностях: {unique_chars_dst}")

Количество уникальных символов в целевых последовательностях: 201
Уникальные символы в целевых последовательностях: {'│', 'X', 'ο', 'B', 'v', 'x', '\x92', 'þ', 'İ', 'ð', '³', '′', '°', ';', '\x97', '£', '`', 'Ý', '\x9d', 'Ι', 'ï', 'ş', ':', 'ư', 'z', 'O', 'd', '-', '@', '─', 'Ë', 'æ', 'I', 'M', '=', 'S', 'ñ', '♪', '鈥', '“', 'É', 's', '#', '¶', '\xa0', 'a', 'Μ', 'á', 'o', '"', 'Ð', 'b', 'é', 'ý', '4', 'ρ', '7', 'ø', 'Ü', 'õ', 'ì', 'Ö', '/', 'N', '0', 't', 'Y', 'L', 'Z', 'i', '¯', 'U', '\u202d', 'œ', '\\', '—', 'r', '”', 'ı', 'Â', 'ò', 'E', '&', 'š', 'P', '^', 'у', 'ă', '}', '1', 'Ν', 'ç', 'ë', 'Ο', '3', 'ú', 'ó', 'å', 'Β', '‚', '檛', '\x8d', 'ü', '¿', 'Æ', 'ª', '?', '5', 'à', '§', 'ē', 'À', '\x83', 'e', 'H', 'R', '9', 'Á', '8', 'ä', 'C', '6', 'ê', 'p', "'", '¡', 'Ż', '~', 'W', 'fl', 'G', 'f', 'l', 'â', 'Τ', '_', 'è', 'g', 'ν', '€', 'ã', '(', '♫', 'Ã', 'J', '″', '\x99', 'y', 'q', 'c', 'D', '{', 'Ä', 'h', '2', '¤', ')', '™', 'F', 'A', '˜', '.', ',', '●', 'Η', 'w', 'n', 'm', 'í', 'V', '\x90'

Посмотрим на максимальную длинну строк для каждой части датасета:

In [9]:
max_lengths = {}

for split in dataset:
    src_lengths = [len(item["src"]) for item in dataset[split] if item["src"] is not None]
    dst_lengths = [len(item["dst"]) for item in dataset[split] if item["dst"] is not None]
    
    max_src_length = max(src_lengths) if src_lengths else 0
    max_dst_length = max(dst_lengths) if dst_lengths else 0
    
    max_lengths[split] = {"src": max_src_length, "dst": max_dst_length}

print(max_lengths)

{'train': {'src': 321, 'dst': 395}, 'validation': {'src': 504, 'dst': 352}, 'test': {'src': 458, 'dst': 0}}


Самая длинная строка присутствующая в датасете имеет длинну 504 символа.

## Подготовка данных

### Токенизатор

#### Нормализация, размер словаря

Нормализация включает в себя некоторую общую очистку, например, удаление ненужных пробельных символов, понижение регистра и/или удаление ударений. Посмотрим как выполняется нормализация токенизатором выбранной нами модели:

In [10]:
from transformers import AutoTokenizer
example_num = 13

tokenizer = AutoTokenizer.from_pretrained(model_checkpoint)



# Проверка нормализации текста
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

print("Оригинальный текст:", text)
print("Нормализованный текст:", tokenizer.backend_tokenizer.normalizer.normalize_str(text))
print("Перевод:", translate)

# Проверка, является ли токенизатор быстрым
print("Токенизатор быстрый:", tokenizer.is_fast)

# Узнаем текущий размер словаря токенизатора
current_vocab_size = tokenizer.vocab_size
print(f"Текущий размер словаря токенизатора: {current_vocab_size}")


Оригинальный текст: ▲◳◈▦▴◳'◈◪ ▽◠◒◠◳◠▦ 45 ▽◠◒▪▦◈◠▨◗ ◝◠◳◠▦ ◙◠◚◫◞■ ◧◐▱▾▦◨▦ ◕▴▱▴▼▴▨ ◚◠◠▫ ◪◈◪▦ ◚◪ ▴▫◓◠◍◬▦◈◠▨◫ ▷◪◓▨▴◞◗ ◕◭▱▩◎◞▴▫◪▦ ◀◗◓ ◠◒◉▪ ◂▱◈▾◐▾▦▾ ◞◇◳▱▩▽◂◓▵
Нормализованный текст: ▲◳◈▦▴◳'◈◪ ▽◠◒◠◳◠▦ 45 ▽◠◒▪▦◈◠▨◗ ◝◠◳◠▦ ◙◠◚◫◞■ ◧◐▱▾▦◨▦ ◕▴▱▴▼▴▨ ◚◠◠▫ ◪◈◪▦ ◚◪ ▴▫◓◠◍◬▦◈◠▨◫ ▷◪◓▨▴◞◗ ◕◭▱▩◎◞▴▫◪▦ ◀◗◓ ◠◒◉▪ ◂▱◈▾◐▾▦▾ ◞◇◳▱▩▽◂◓▵
Перевод: Davis, 45, who lives in Leadney, said her son was a great cook and had a charming smile.
Токенизатор быстрый: True
Текущий размер словаря токенизатора: 32100


Судя по выводу текст подвергается минимальной нормализации, значит проблем с потерей символов быть не должно. Размер предобученного словаря токенизатора для выбранной модели равен 32100 токен. 

#### Размер контекстного окна

In [11]:
# Получим максимальную длину последовательности (размер контекстного окна) модели 
max_length = tokenizer.model_max_length

print(f"Максимальная длина последовательности модели: {max_length}")

Максимальная длина последовательности модели: 512


Максимальная длинна контекста для выбранной модели равна 512 токенов.

> ***Длина контекста (или контекстное окно)*** это максимальное количеству токенов, которые модель может обрабатывать одновременно.

#### Обучение нового токенизатора

Так как язык зетан является вымешленным (новым), необходимо обучить новый токенизатор:

In [12]:
def batch_iterator(batch_size=10):
    for split in dataset:
        batch_size = batch_size // 2
        for i in range(0, len(dataset[split]), batch_size):
            src_list = dataset[split][i : i + batch_size]["src"]
            dst_list = dataset[split][i : i + batch_size]["dst"]
            yield [str(src) for src in src_list] + [str(dst) for dst in dst_list]


vocab_size = current_vocab_size  # Можно начать с 32000 и экспериментировать

# Обучение токенизатора
new_tokenizer = tokenizer.train_new_from_iterator(batch_iterator(batch_size=1000), vocab_size=vocab_size)
print("Токенизатор обучен")




Токенизатор обучен


Проверим работу нового токенизатора на нескольких примерах исходного и целевого текста:

In [13]:
example_num = 1

# Получаем строки из датасета
text = dataset['validation'][example_num]['src']
translate = dataset['validation'][example_num]['dst']

# Токенизируем текст и перевод, выводим результат
print(text)
print(new_tokenizer(text), end="\n\n")

print(translate)
print(new_tokenizer(translate), end="\n\n")

print(new_tokenizer.vocab)

◤◪▦◫ ▨◠▦◞▴◓ ◠◒▪◞▪■ ◀◠◐▪◒◬▨▱▪▨ ◞◫◞▫◪◎◫▦▴ ▨◣▫◭ ▦◫◳◪▫▱◗ ▷▩▼◓◪▱▴◓◫ "◕◣◓◎▴▽◫" ◇◐◓◪▫▴◀◫▱◗◓
{'input_ids': [10743, 18257, 16672, 11029, 14803, 1290, 585, 11060, 1372, 1907, 13501, 11473, 16422, 517, 267, 460, 2572, 7108, 179, 10515, 3414, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

A new cancer vaccine may teach the immune system to "see" abnormal cells
{'input_ids': [251, 543, 17698, 24831, 742, 4487, 107, 22203, 3659, 109, 267, 123, 3220, 179, 111, 562, 471, 10123, 532, 17617, 1], 'attention_mask': [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]}

{'▁▨◠◈◠◓◬◳▱◠': 18739, '▁◠◈◬': 3359, '▁prop': 17740, '◎◠▽◠▼◠◐◬▦◬': 24116, '▁○◉▪▨◉◠◞▪': 29910, "▁◦◍◕◠▦◫◞▫◠▦'": 31694, '2': 1962, '▁▨◠◎▻◠▦▽◠': 20309, 'we': 3324, '▁◓▾▷▾': 24300, '◎◠◞◬▦◬': 5173, '▁i': 1152, '▁father': 942, '▁◠▦▦▴◞◫': 17326, '▁◠▦◠▷▫◠◓': 5118, '▁◝◗▱◫◳◧◓': 17288, '00': 2341, 'Go': 22176, '◞▴▦◫▢': 8206, '▁▷◠▫▫◠': 4728, '▁◉◂◓◠▻': 23274, '▁important.': 5341, '▁faculty': 28146, '▁

Сохраним полученный новый токенизатор локально: 

In [14]:
new_tokenizer.save_pretrained("./Model/new_tokenizer")

('./Model/new_tokenizer/tokenizer_config.json',
 './Model/new_tokenizer/special_tokens_map.json',
 './Model/new_tokenizer/tokenizer.json')

### Подготовка данных для обучения модели

Загрузим обученный ранее токенизатор:


In [15]:
tokenizer = AutoTokenizer.from_pretrained("./Model/new_tokenizer")

Проверим токенизатор на паре предложений из тренировочной части датасета:

In [18]:
dataset['train'][example_num]['src']

'◝◨◓◠◈◠▨◗ ▷◪◓▨◪◞◫ ◳◠▽◠ ◀◬◓◠▨▪◓▵'

In [22]:
example_num = 7
test_examples = [dataset['train'][example_num]['src'], dataset['train'][example_num]['dst']]
tokenizer(test_examples)

{'input_ids': [[1057, 474, 3304, 152, 572, 288, 20158, 103, 1], [10065, 3247, 123, 107, 8809, 2014, 147, 1983, 123, 4151, 309, 1]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]]}

Для разных моделей может понадобиться разный подход в настройке перевода. Например для моделей семества `google-t5` необходим специальный преффикс. Реализуем это:

In [23]:
print(model_checkpoint)

if model_checkpoint in ["google-t5/t5-small", "google-t5/t5-base", "google-t5/t5-larg", "google-t5/t5-3b", "google-t5/t5-11b"]:
    print("This is a T5 model")
    prefix = "translate Zetan to English: "
else:
    print("This is not a T5 model")
    prefix = ""

google-t5/t5-small
This is a T5 model


Прежде чем передать в модель тренировочнуй часть датасета необходимо его предварительно обработать. Напишем функцию которая подготовит датасет в которой:

- примеры из датасета будут подаваться на вход `tokenizer` с аргументом `truncation=True`. Это гарантирует, что входные данные, длина которых превышает ту, которую может обработать выбранная модель, будут усечены до максимальной длины, принимаемой моделью.

- Дополнение коротких последовательностей (Padding) будет реализованно позже (в коллаторе данных).

In [ ]:
# Установка максимальной длины входной и выходной последовательности
max_input_length = max_length
max_target_length = max_length

source_lang = "zetan"
target_lang = "en"

def preprocess_function(examples):

    inputs = [prefix + ex for ex in examples['src']]
    targets = [ex for ex in examples['dst'] if ex is not None]

    # Токенизируем входные данные и цели
    model_inputs = tokenizer(inputs, max_length=max_input_length, truncation=True)

    with tokenizer.as_target_tokenizer():
        labels = tokenizer(targets, max_length=max_target_length, truncation=True)

    model_inputs["labels"] = labels["input_ids"]
    return model_inputs

Проверим функцию на нескольких примерах из датасета:

In [89]:
preprocess_function(dataset['train'][:2])

{'input_ids': [[24968, 9098, 6621, 109, 5326, 355, 10912, 9539, 1285, 3017, 103, 1], [24968, 9098, 6621, 109, 5326, 355, 19463, 1035, 1693, 1513, 261, 163, 999, 3413, 1817, 24866, 713, 1453, 18113, 1]], 'attention_mask': [[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1], [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]], 'labels': [[114, 122, 32007, 104, 1], [213, 243, 240, 109, 7905, 200, 21353, 123, 118, 107, 3804, 115, 107, 1748, 117, 1866, 202, 107, 4570, 115, 18107, 104, 1]]}

Загрузим датасет без тестовой части, так как в ней нет целей для предсказания. Используем метод `map` чтобы применить функцию к оставщейся части датасета:

In [96]:
data_files = {
    "train": "train.jsonl",
    "validation": "val.jsonl",
}

features = Features({
    'src': Value('string'),
    'dst': Value('string'),
})

dataset = load_dataset("json", data_files=data_files, features=features)

dataset

DatasetDict({
    train: Dataset({
        features: ['src', 'dst'],
        num_rows: 300000
    })
    validation: Dataset({
        features: ['src', 'dst'],
        num_rows: 500
    })
})

In [97]:
# Применение функции предобработки к датасету
tokenized_datasets = dataset.map(preprocess_function, batched=True)

Map:   0%|          | 0/300000 [00:00<?, ? examples/s]

/usr/local/lib/python3.12/dist-packages/transformers/tokenization_utils_base.py:4114: UserWarning: `as_target_tokenizer` is deprecated and will be removed in v5 of Transformers. You can tokenize your labels by using the argument `text_target` of the regular `__call__` method (either in the same call as your input texts if you use the same keyword arguments, or in a separate call.
  warnings.warn(


Map:   0%|          | 0/500 [00:00<?, ? examples/s]

# Обучение модели

Загрузим необученную модель (модель со случайно сгенерированными весами):

In [35]:
from transformers import AutoConfig, T5ForConditionalGeneration

config = AutoConfig.from_pretrained(model_checkpoint)
model = T5ForConditionalGeneration(config)

config.json:   0%|          | 0.00/1.21k [00:00<?, ?B/s]

# Дополнительные материалы

По основной задаче:

- [Training your own tokenizer from scratch](https://github.com/huggingface/notebooks/blob/main/examples/tokenizer_training.ipynb)
- [Translation fine-tuning example](https://github.com/huggingface/notebooks/blob/main/examples/translation.ipynb)

По проблемам с контейнером:

- [Jupyter notebook executions turn grey in Visual studio code](https://stackoverflow.com/questions/73481249/jupyter-notebook-executions-turn-grey-in-visual-studio-code)

# Тестовая часть

Используется для отладки кода и проведения экспериментов.